### PR TITLE
make direct call to V2 handlePrivateTx from dag

### PR DESCRIPTION
**Pull request by woutslakhorst**
handles problem from #711

It calls the V2 protocol from a DAG subscriber to fetch a private TX.
As you can tell, the subscribers setup is dodgy and will have to be replaced with a proper event mechanism.
This was the purpose of Nats but that part is missing proper design for now.

- [x] cherry-pick #739 
- [x] cherry-pick #740 (splitted callbacks)
- [x] cherry-pick #744 (did doc creation)
- [x] cherry-pick #741 (publisher retry change)

### CODE REVIEW

**Comment by dmeijboom** [comment]
I'll postpone reviewing this until #708 is merged 😅

File: vcr/ambassador.go [hunk 1]
```diff
@@ -59,6 +59,11 @@ func (n ambassador) Configure() {
 // The VCR is used to verify the contents of the credential.
 // payload should be a json encoded vc.VerifiableCredential
 func (n ambassador) vcCallback(tx dag.Transaction, payload []byte) error {
+	if payload == nil {
+		// ignore first call, only act on second call (with payload)
+		return nil
+	}
```
Comment (dmeijboom) [inline vcr/ambassador.go:65]: Why is this required?

File: vdr/ambassador.go [hunk 2]
```diff
@@ -83,6 +83,11 @@ var thumbprintAlg = crypto.SHA256
 // The rules are based on the Nuts RFC006
 // payload should be a json encoded did.document
 func (n *ambassador) callback(tx dag.Transaction, payload []byte) error {
+	if payload == nil {
+		// ignore first call, only act on second call (with payload)
+		return nil
+	}
```
Comment (dmeijboom) [inline vdr/ambassador.go:89]: Why is this required?

**Review by dmeijboom — CHANGES_REQUESTED** [review]

**Comment by woutslakhorst** [comment]
The e2e tests found a problem..... and they are correct. More things to think about....

File: vcr/ambassador.go (hunk 1 above)
Comment (woutslakhorst) [reply vcr/ambassador.go]: has been removed

File: vdr/ambassador.go (hunk 2 above)
Comment (woutslakhorst) [reply vdr/ambassador.go]: has been removed

**Review by dmeijboom — APPROVED** [review]
LGTM

File: test/predicates.go [hunk 3]
```diff
@@ -41,6 +42,23 @@ func WaitFor(t *testing.T, p Predicate, timeout time.Duration, message string, m
 			assert.Fail(t, fmt.Sprintf(message, msgArgs...))
 			return false
 		}
-		time.Sleep(50 * time.Millisecond)
+		time.Sleep(10 * time.Millisecond)
+	}
+}
+
+func WaitForNoFail(t *testing.T, p Predicate, timeout time.Duration) bool {
```
Comment (reinkrul) [inline test/predicates.go:49]: Let's not wait too long ;-)

**Review by reinkrul — APPROVED** [review]
Much better now.

File: network/dag/publisher.go [hunk 4]
```diff
@@ -77,7 +75,7 @@ func (s *replayingDAGPublisher) transactionAdded(ctx context.Context, transactio
 	s.emitEvent(TransactionAddedEvent, tx, nil)
 
 	// Received new transaction, add it to the subscription walker resume list, so it resumes from this transaction
-	// when the payload is received.
+	// when the payload is received. This block so it may only be added for TX where we know the payload will come.
```
Comment (reinkrul) [inline network/dag/publisher.go:78]:
```suggestion
	// when the payload is received. This blocks so it may only be added for TX where we know the payload will come.
```

**Review by reinkrul — APPROVED** [review]

File: network/dag/publisher.go [hunk 5]
```diff
@@ -134,6 +135,21 @@ func (s *replayingDAGPublisher) publish(ctx context.Context) {
...
 
+func remove(l *list.List, ref hash.SHA256Hash) {
+	current := l.Front()
+
+	for {
+		if current == nil {
+			return
+		}
+		if current.Value.(hash.SHA256Hash).Equals(ref) {
+			l.Remove(current)
```
Comment (reinkrul) [inline network/dag/publisher.go:146]: could there be duplicate resume points? If so, we could loop over the entire list, trying to remove duplicates

Comment (woutslakhorst) [reply network/dag/publisher.go:147]: Very unlikely the transactionAdded callback is called twice. Even so, doesn't hurt to remove the `return` here.

**Comment by dmeijboom** [comment]
@woutslakhorst @reinkrul Looks like this works with the latest Demo EHR. Also verified it actually uses a private transaction. Setup with three nodes where node 1 performed an e-transfer to node 2 and node 3 wasn't able to decrypt the PAL header. Nice work 🎉
